extension/js/content_scripts/webmail/setup-webmail-content-script.ts

```diff
@@ -238,6 +239,11 @@ export const contentScriptSetupIfVacant = async (webmailSpecific: WebmailSpecifi
     BrowserMsg.addListener('show_attachment_preview', async ({ iframeUrl }: Bm.ShowAttachmentPreview) => {
       await Ui.modal.attachmentPreview(iframeUrl);
     });
+    BrowserMsg.addListener('show_warning_for_attachment_download', async ({ message, attachment }: Bm.ShowWarningForAttachmentDownload) => {
+      if (await Ui.modal.confirm(message)) {
+        Browser.saveToDownloads(attachment);
```

Comment:
For some unknown reason, this doesn't work. I think I need to put the confirmation modal into an `iframe` similar to what `showAttachmentPreview` does. Then add a test later for live Gmail and web pages.

Reply:
the `attachment` variable inside `show_warning_for_attachment_download` (webmail) event listener shows limited property/fields -- compared to fields to attachment.js (common/core). It is the reason why `saveToDownloads()` fails since attachment.getData() is not present.
![image](https://user-images.githubusercontent.com/46025304/212830401-0bd86109-a937-417b-8d19-9ea5a42902b1.png).

I check the typescript definition in `browser_msg.ts` for my background message event handler and it is correct.

@rrrooommmaaa @ioanmo226 do you have any suggestions? Thanks in advance.

Reply:
You are right. Attachment variable has limited fields.
Please check this article.
https://stackoverflow.com/questions/32814553/sending-object-with-functions-via-chrome-runtime-sendmessage#:~:text=You%20can%27t%20pass,then%20utilize%20it.

Reply:
Functions get stripped away when sending (serializing) the object

Reply:
Maybe we can copy-paste this getData function logic into `Browser.saveToDownloads` function?
@rrrooommmaaa What's your suggestion?

Reply:
> Maybe we can copy-paste this getData function logic into Browser.saveToDownloads function?

I can also see it to be a possibility. Will wait for Roman's suggestion. 

Thank you very much Ioan and Roman.

Reply:
As I understand, there is no need to call `Browser.saveToDownloads` from the background, only the confirmation?
So we could add a listener to show a confirmation and return `boolean`, that is, instead of `show_warning_for_attachment_download` simply `show_confirmation` with the message to show, without passing the attachment.
then we can get the result with `sendAwait` and do the `Browser.saveToDownloads` call from the calling thread. Would that work?

Reply:
Sounds okay to me.

Reply:
oh this wouldn't work as the response for the message is required immediately

Reply:
1) One option is to implement a callback similar to what we have for `passphrase_dialog`.
This is a bit cumbersome but useful, so that we can easily show confirmation popups from other places when needed.
2) Another option is that we can split Browser.saveToDownloads into 2 methods. One method extracts data as Uint8Array and attachment's name and type and forms this intermediate object which we're able to send to background. And the other method takes this intermediate object and does the remaining work.
But I'm not happy with scattering `Browser.saveToDownloads` (or whatever the new sub-method may be called) across several modules -- this potentially creates a mess.

It's better to implement confirmation separately and do as less `Browser.saveToDownloads` calls as possible (or do these calls from a less amount of modules as possible).
3) So the third option would be to route ALL file download handling via the background (with the intermediate object), and do the file extension check in the background code.
Each option has its pros and cons.

Reply:
@rrrooommmaaa I think I get your point about showing the prompt and waiting for the response from a browser message. I tried to make changes according to the description like simplification and try reading the browser message response in the right place.

I'm not getting a response after trying to return the data from `Ui.modal.confirm()`.

![image](https://user-images.githubusercontent.com/46025304/214010306-cb9dd06e-11a4-4827-91c8-9c7496e8ca67.png)

```
Error downloading file - Error: BrowserMsg(show_confirmation) sendRawResponse::Error: BrowserMsg.listen error: handler "show_confirmation" not set.
```

I'm getting this error but I can't identify where could I go wrong since the `show_confirmation` called in the setup-webmail and inbox.ts -- the pop-up confirmation appears.

We can use the test here: https://mail.google.com/mail/u/flowcrypt.compatibility@gmail.com/#inbox/FMfcgzGrcFnCsQsNGKRQdVdHBbQKgRhL

Reply:
Right, the messaging system is a bit complicated
-- as you can see in comments to `private static shouldRelayMsgToOtherPage`, the message can get to `setup-webmail-content-script.ts` by reaching the background page (relayed inside `bgListen()`), so the response callback is lost. (For now) we can only receive responses when sending messages to the background handler (with `BrowserMsg.sendAwait(undefined, ...`). Can you try and install the handler to the `background_page.ts` (instead of `inbox.ts` and `setup-webmail-content-script.ts`)? I vaguely remember that it is possible to show a modal from a background page handler, but not sure.

Reply:
@rrrooommmaaa I tried to move the handlers to background pages and they don't seem to work. Although it is capable of returning plain data but not UI modal I think?

Reply:
> Although it is capable of returning plain data like strings/boolean but not UI modal I think?

Maybe. Please add a test message with an executable file and a test to decrypt it. They'll be useful for debugging.